lib/view_model/active_record/update_data.rb

```diff
@@ -24,45 +18,325 @@ def self.for_type(type)
       end
     end
 
-    class Append < self
+    class Append
       NAME = 'append'
       attr_accessor :before, :after
-      def self.schema
-        UpdateData::Schemas::APPEND_ACTION
+      attr_reader :contents
+      def initialize(contents)
+        @contents = contents
       end
     end
 
-    class Remove < self
+    class Update
+      NAME = 'update'
+      attr_reader :contents
+      def initialize(contents)
+        @contents = contents
+      end
+    end
+
+    class Remove
       NAME = 'remove'
-      def self.schema
-        UpdateData::Schemas::REMOVE_ACTION
+      attr_reader :removed_vm_refs
+
+      def initialize(removed_vm_refs)
+        @removed_vm_refs = removed_vm_refs
       end
     end
+  end
 
-    class Update < self
-      NAME = 'update'
-      def self.schema
-        UpdateData::Schemas::UPDATE_ACTION
+  # Parser for collection updates. Collection updates have a regular
+  # structure, but vary based on the contents. Updates to a direct,
+  # owned collection recurse deeply and create a tree of UpdateDatas,
+  # while through, shared collections collected reference strings.
+  class AbstractCollectionUpdateParser # private
+
+    def initialize(association_data, blame_reference, valid_reference_keys)
+      @association_data     = association_data
+      @blame_reference      = blame_reference
+      @valid_reference_keys = valid_reference_keys
+    end
+
+    def parse(value)
+      case value
+      when Array
+        new_replace(parse_contents(value))
+
+      when Hash
+        ViewModel::Schemas.verify_schema!(action_schema, value)
```

Comment:
It's probably a `functional_update_schema`, i.e., the whole update, which has a list of actions, and its type, etc.